File: solutions/python/easy/p0121_best_time_to_buy_and_sell_stock.py
"""LeetCode 121: Best Time to Buy and Sell Stock.

Problem: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
"""

from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        """Track the lowest price seen before each possible sale.

        Time: O(n)
        Space: O(1)
        """
        lowest_price = float("inf")
        best_profit = 0
        for price in prices:
            lowest_price = min(lowest_price, price)
            best_profit = max(best_profit, price - lowest_price)
        return best_profit

File: solutions/java/easy/P0121_BestTimeToBuyAndSellStock.java
/**
 * LeetCode 121: Best Time to Buy and Sell Stock.
 *
 * Problem: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
 */
public class P0121_BestTimeToBuyAndSellStock {
    public int maxProfit(int[] prices) {
        int minPrice = prices[0];
        int maxProfit = 0;
        for (int i = 1;i < prices.length;i++){
            int profit = prices[i] - minPrice;
            if (profit > maxProfit) {
                maxProfit = profit;
            }
            if (prices[i] < minPrice) {
                minPrice = prices[i];
            }
        }
        return maxProfit;
    }

    public static void main(String[] args) {
        P0121_BestTimeToBuyAndSellStock solution = new P0121_BestTimeToBuyAndSellStock();
        System.out.println(solution.maxProfit(new int[] {7, 1, 5, 3, 6, 4})); // Expected: 5
        System.out.println(solution.maxProfit(new int[] {7, 6, 4, 3, 1}));    // Expected: 0
    }
}
